File: solutions/haskell/dominoes/1/src/Dominoes.hs
-- With deepest reverence to Shree DR.MDD ⚜️

module Dominoes (chain) where

import Control.Monad.State (evalState, MonadState(get, put), StateT)
import Data.Maybe (listToMaybe)
import Data.Tuple (swap)
import qualified Control.Monad.Identity as ID
import Data.List (delete)

type Tile = (Int, Int)
type Layout = [Tile]
type Pathway = ([Tile], Layout)
type Trailbook = [Pathway]

chain :: [(Int, Int)] -> Maybe [(Int, Int)]
chain [] = Just []
chain (t:ts) = snd <$> result
  where
    initialPaths :: Trailbook
    initialPaths = [(ts, [t])]
    finalPaths :: Trailbook
    finalPaths = evalState findPaths initialPaths
    result :: Maybe Pathway
    result = listToMaybe finalPaths

findPaths :: StateT Trailbook ID.Identity Trailbook
findPaths = do
    currSet <- get
    if any (null . fst) currSet
    then return $ filter verifyLoop currSet
    else do
        let expanded = branchOut currSet
        if null expanded
        then return []
        else do
            put expanded
            findPaths

verifyLoop :: Pathway -> Bool
verifyLoop (_, seqn) = fst (head seqn) == snd (last seqn)

branchOut :: Trailbook -> Trailbook
branchOut = concatMap stepAhead

stepAhead :: Pathway -> Trailbook
stepAhead (pile, layout) = concatMap (link layout pile) pile

link :: Layout -> [Tile] -> Tile -> Trailbook
link layout avail tile = validCombos
  where
    remain = delete tile avail
    possibles = map (:layout) $ filter matchesEnd [tile, swap tile]
    validCombos = map (\x -> (remain, x)) possibles
    matchesEnd t = fst (head layout) == snd t

File: solutions/elixir/dominoes/1/lib/dominoes.ex
# For Shree DR.MDD

defmodule Dominoes do
  @type tile :: {1..6, 1..6}

  @spec chain?([tile] | []) :: boolean
  def chain?([]), do: true

  def chain?([start_tile | rest]) do
    traverse_chain?({start_tile, start_tile}, rest)
  end

  defp traverse_chain?({{a, _}, {_, a}}, []), do: true
  defp traverse_chain?(_, []), do: false

  defp traverse_chain?({first, {_, end_val}}, tiles) do
    Enum.any?(tiles, fn
      next = {^end_val, _} -> traverse_chain?({first, next}, List.delete(tiles, next))
      next = {x, ^end_val} -> traverse_chain?({first, {end_val, x}}, List.delete(tiles, next))
      _ -> false
    end)
  end
end
